File: 128. 最长连续序列.py
class Solution(object):
    def longestConsecutive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums_set = set(nums)
        length_list = [0]
        for num in nums_set:
            if num - 1 not in nums_set:  # 如果num不是序列第一位，就从它开始往后找
                length = 1
                while num + 1 in nums_set:
                    length += 1
                    num += 1
                length_list.append(length)
        return max(length_list)


s = Solution()
print(s.longestConsecutive([100, 4, 200, 1, 3, 2]))

File: 128. 最长连续序列.go
package main

func longestConsecutive(nums []int) int {
	// 1. 把所有数字放进集合
	// 2. 遍历集合中的每个数字
	// 3. 如果 num-1 存在，说明 num 不是起点，跳过
	// 4. 如果 num-1 不存在，从 num 开始不断检查 num+1、num+2……
	// 5. 记录最长长度

	var res int

	s := make(map[int]struct{})
	for _, num := range nums {
		s[num] = struct{}{}
	}

	for num := range s {
		if _, ok := s[num-1]; ok {
			continue
		}

		currNum := num
		for {
			if _, ok := s[currNum+1]; ok {
				currNum++
				continue
			}
			break
		}
		length := currNum - num + 1
		if length > res {
			res = length
		}
	}
	return res
}
